updater/updater_test.go

```diff
@@ -0,0 +1,179 @@
+	assert.NoError(t, err)
+
+	u := newUpdater(a, tApp.ID, tGroup.Track, "instance-0001", "0.2.0")
+	info, err := u.CheckForUpdates(context.TODO())
+	assert.NoError(t, err)
+	assert.False(t, info.HasUpdate)
+	assert.Equal(t, "", info.GetVersion())
+
+	newPkg, _ := a.AddPackage(&api.Package{Type: api.PkgTypeOther, URL: "http://sample.url/pkg", Version: "0.3.0", ApplicationID: tApp.ID, Arch: api.ArchAMD64, Filename: null.StringFrom("updatefile.txt")})
+	tChannel.PackageID = null.StringFrom(newPkg.ID)
+	err = a.UpdateChannel(tChannel)
+	assert.NoError(t, err)
+
+	info, err = u.CheckForUpdates(context.TODO())
+	assert.NoError(t, err)
+	assert.True(t, info.HasUpdate)
+
+	version := info.GetVersion()
+	assert.Equal(t, "0.3.0", version)
+
+	urls := info.GetURLs()
+	assert.NotNil(t, urls)
+	assert.Equal(t, 1, len(urls))
+	assert.Equal(t, urls[len(urls)-1], info.GetURL())
+	assert.Equal(t, "http://sample.url/pkg", info.GetURL())
+
+	pkg := info.GetPackage()
+	assert.NotNil(t, pkg)
+	assert.Equal(t, "updatefile.txt", pkg.Name)
+}
+
+type updateTestHandler struct {
+	fetchUpdateResult error
+	applyUpdateResult error
+}
+
+func (u updateTestHandler) FetchUpdate(ctx context.Context, info *UpdateInfo) error {
+	// Nothing to do
+	return u.fetchUpdateResult
+}
+
+func (u updateTestHandler) ApplyUpdate(ctx context.Context, info *UpdateInfo) error {
+	// Nothing to do
+	return u.applyUpdateResult
+}
+
+func TestTryUpdate(t *testing.T) {
+	a := newForTest(t)
+	defer a.Close()
+
+	tTeam, _ := a.AddTeam(&api.Team{Name: "test_team"})
+	tApp, _ := a.AddApp(&api.Application{Name: "io.phony.App", TeamID: tTeam.ID})
+	tPkg, _ := a.AddPackage(&api.Package{Type: api.PkgTypeOther, URL: "http://sample.url/pkg", Version: "0.4.0", ApplicationID: tApp.ID, Arch: api.ArchAMD64})
+	tChannel, _ := a.AddChannel(&api.Channel{Name: "channel1", Color: "blue", ApplicationID: tApp.ID, PackageID: null.StringFrom(tPkg.ID), Arch: api.ArchAMD64})
+	tGroup, err := a.AddGroup(&api.Group{Name: "group1", ApplicationID: tApp.ID, ChannelID: null.StringFrom(tChannel.ID), PolicyUpdatesEnabled: true, PolicySafeMode: false, PolicyPeriodInterval: "15 minutes", PolicyMaxUpdatesPerPeriod: 10, PolicyUpdateTimeout: "60 minutes", Track: "stable"})
+	assert.NoError(t, err)
+
+	oldVersion := "0.2.0"
+
+	u := newUpdater(a, tApp.ID, tGroup.Track, "instance-0001", "0.2.0")
+	assert.Equal(t, oldVersion, u.instanceVersion)
+
+	// Error when fetching update
+	err = u.TryUpdate(context.TODO(), &updateTestHandler{
+		errors.New("something went wrong fetching the update"),
+		nil,
+	})
+	assert.Error(t, err)
+	assert.Equal(t, oldVersion, u.instanceVersion)
+
+	// Error when applying update
+	err = u.TryUpdate(context.TODO(), &updateTestHandler{
+		nil,
+		errors.New("something went wrong applying the update"),
+	})
```

Comment:
IIRC one should not rely on fields order in struct, as it may silently break code when those are swapped (e.g. as part of memory-layout optimization).
```suggestion
	err = u.TryUpdate(context.TODO(), &updateTestHandler{
		applyUpdateResult: errors.New("something went wrong applying the update"),
	})
```